content/docs/guides/crossguard/configuration.md

```diff
@@ -0,0 +1,132 @@
+linktitle: Configuration
+weight: 2
+
+menu:
+  userguides:
+    parent: crossguard
+---
+<!-- markdownlint-disable ul code -->
+{{% crossguard-preview %}}
+
+## Overview
+
+Configuration allows you to author flexible Policy Packs that can be reused across your entire organization. As an organization administrator, you can use a single Policy Pack yet vary configuration (e.g. enforcement level, the allowed instance types, cost allowances, etc.) from one Policy Group to the next. For example, you may have a Policy Group for your non-production stacks that allow smaller instance types, while your Production Policy Group allows for use of large instance types.
+
+Configuration schema is defined per policy, and then the actual configuration can be set via a form in the Pulumi Console or using a JSON file.
+
+Configuration is currently supported for the TypeScript and JavaScript Policy SDKs. Python support for configuration is [coming soon](https://github.com/pulumi/pulumi-policy/issues/210).
+
+## Writing Configurable Policy Packs
+
+### Enforcement Level
+
+The Policy SDK will automatically make the enforcement levels of all policies configurable. You can configure the enforcement level for all policies in a Policy Pack or configure it for each individual policy level. The top level enforcement level under `all` will override the existing enforcement levels of all policies in the Policy Pack. A configured enforcement level for a specific policy will override the policy's existing enforcement level as well as the top level configured enforcement level.
+
+In the example configuration below, all policies in the Policy Pack would be `disabled` with the exception of the `a-policy` policy, which would be `mandatory`.
+
+```json
+{
+    "all": {
+        "enforcementLevel": "disabled",
+    },
+    "a-policy": {
+         "enforcementLevel": "mandatory",
+    }
+}
+```
+
+### Custom Configuration
+
+Policy authors can denote the schema for a particular Policy's configuration using the `configSchema` field. The configuration schema uses [JSON Schema](https://json-schema.org/) to describe the acceptable configuration for a Policy.
+
+The below example shows a  `ResourceValidationPolicy` that has a required `message` field, which must be between five and fifty characters. Then, from within the `validateResource` function the configuration value can be accessed using `args.getConfig`.
+
+```typescript
+const examplePolicy: ResourceValidationPolicy = {
+    name: "example-policy-with-schema",
+    ...
+    configSchema: {
+        properties: {
+            message: {
+                type: "string",
+                minLength: 5,
+                maxLength: 50,
+                default: "This is a configurable message."
+            }
+        },
+        required: ["message"],
+    },
+    validateResource: validateTypedResource(aws.s3.Bucket, (_, args, reportViolation) => {
+        const config = args.getConfig<{ message?: string }>();
```

Comment:
Since you've specified `required: ["message"]`, this can be typed as:

```suggestion
        const config = args.getConfig<{ message: string }>();
```

...without the `?` after `message`. Message will always be set since it was specified as being required.


However, I wonder if we should lead with an example that doesn't use a required property, and have some verbiage about how properties are optional by default. And show the use of the `?` in the type specified with the `getConfig<T>` call, and then inside the validation handler, show how to check if it is set, but otherwise still have the policy be usable without the config flag being set. Because we generally recommend not using required config properties -- that just adds friction in enabling/adopting the policy. It's generally better to provide policies that can run without any config, but can change behavior or opt-in to more specific checking if certain config properties _are_ set.

For example, a policy that checks encryption is enabled (that can do that check without any config), that has an optional config value for an "id of the encryption key" that, when specified, _further_ enforces that that specific key is being used for the encryption.

And then have a separate paragraph that mentions, how in some cases a policy by it's very nature necessitates a required property, in which case, here's how you'd do it... How the `?` in the `getConfig<T>` type isn't necessary because the system will enforce it being set, etc.

Reply:
- [ ] lead with non-required example
- [ ] discuss properties being optional by default
- [ ] generally recommend not having required PPs bc it adds friction